File: android/src/main/java/com/MythiCode/camerakit/CameraKitFlutterView.java
package com.MythiCode.camerakit;

import android.Manifest;
import android.content.pm.PackageManager;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;

import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.platform.PlatformView;

public class CameraKitFlutterView implements PlatformView, MethodChannel.MethodCallHandler, FlutterMethodListener {

    private static final int REQUEST_CAMERA_PERMISSION = 10001;
    private final MethodChannel channel;
    private final ActivityPluginBinding activityPluginBinding;
    private CameraBaseView cameraView;

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull final MethodChannel.Result result) {
        switch (call.method) {
            case "requestPermission":
                if (ActivityCompat.checkSelfPermission(activityPluginBinding.getActivity(), Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                    ActivityCompat.requestPermissions(activityPluginBinding.getActivity(), new String[]{Manifest.permission.CAMERA}, REQUEST_CAMERA_PERMISSION);
                    activityPluginBinding.addRequestPermissionsResultListener(new PluginRegistry.RequestPermissionsResultListener() {
                        @Override
                        public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
                            for (int i :
                                    grantResults) {
                                if (i == PackageManager.PERMISSION_DENIED) {
                                    try {
                                        result.success(false);
                                    } catch (Exception ignored) {

                                    }
                                    return false;
                                }
                            }
                            try {
                                result.success(true);
                            } catch (Exception ignored) {

                            }
                            return false;
                        }
                    });
                    return;
                } else {
                    try {
                        result.success(true);
                    } catch (Exception ignored) {

                    }
                }
                break;
            case "initCamera":
                boolean hasBarcodeReader = call.argument("hasBarcodeReader");
                char flashMode = call.argument("flashMode").toString().charAt(0);
                boolean isFillScale = call.argument("isFillScale");
                int barcodeMode = call.argument("barcodeMode");
                int androidCameraMode = call.argument("androidCameraMode");
                int cameraSelector = call.argument("cameraSelector");
                getCameraView().initCamera(hasBarcodeReader, flashMode, isFillScale, barcodeMode, androidCameraMode, cameraSelector);
                break;
            case "resumeCamera":
                getCameraView().resumeCamera();

                break;
            case "pauseCamera":
                getCameraView().pauseCamera();
                break;
            case "takePicture":
                String path = call.argument("path").toString();
                getCameraView().takePicture(path, result);
                break;
            case "changeFlashMode":
                char captureFlashMode = call.argument("flashMode").toString().charAt(0);
                getCameraView().changeFlashMode(captureFlashMode);
                break;
            case "dispose":
                dispose();
                break;
            case "setCameraVisible":
                boolean isCameraVisible = call.argument("isCameraVisible");
                getCameraView().setCameraVisible(isCameraVisible);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private CameraBaseView getCameraView() {
        return cameraView;
    }

    public CameraKitFlutterView(ActivityPluginBinding activityPluginBinding, @NonNull BinaryMessenger messenger, int id) {
        this.channel = new MethodChannel(messenger, "plugins/camera_kit_" + id);
        this.activityPluginBinding = activityPluginBinding;
        this.channel.setMethodCallHandler(this);
        if (getCameraView() == null) {
            cameraView = new CameraBaseView(activityPluginBinding.getActivity(), this);
        }
    }

    @Override
    public View getView() {
        return getCameraView().getView();
    }

    @Override
    public void dispose() {
        if (getCameraView() != null) {
            getCameraView().dispose();
        }
    }

    @Override
    public void onBarcodeRead(String barcode) {
        channel.invokeMethod("onBarcodeRead", barcode);
    }

    @Override
    public void onTakePicture(final MethodChannel.Result result, final String filePath) {
        activityPluginBinding.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(filePath);
            }
        });
    }

    @Override
    public void onTakePictureFailed(final MethodChannel.Result result, final String errorCode, final String errorMessage) {
        activityPluginBinding.getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.error(errorCode, errorMessage, null);
            }
        });
    }
}

File: android/src/main/kotlin/com/MythiCode/camerakit/CameraKitFlutterView.kt
package com.MythiCode.camerakit

import android.Manifest
import android.content.pm.PackageManager
import android.view.View
import androidx.core.app.ActivityCompat
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.platform.PlatformView
import java.util.*

class CameraKitFlutterView(
    private val activityPluginBinding: ActivityPluginBinding,
    messenger: BinaryMessenger,
    id: Int
) : PlatformView, MethodCallHandler, FlutterMethodListener {
    private val channel: MethodChannel = MethodChannel(messenger, "plugins/camera_kit_$id")
    private var cameraView: CameraBaseView? = null
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "requestPermission" -> if (ActivityCompat.checkSelfPermission(
                    activityPluginBinding.activity,
                    Manifest.permission.CAMERA
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    activityPluginBinding.activity,
                    arrayOf(Manifest.permission.CAMERA),
                    REQUEST_CAMERA_PERMISSION
                )
                activityPluginBinding.addRequestPermissionsResultListener { _: Int, _: Array<String?>?, grantResults: IntArray ->
                    for (i in grantResults) {
                        if (i == PackageManager.PERMISSION_DENIED) {
                            try {
                                result.success(false)
                            } catch (ignored: Exception) {
                            }
                            return@addRequestPermissionsResultListener false
                        }
                    }
                    try {
                        result.success(true)
                    } catch (ignored: Exception) {
                    }
                    false
                }
                return
            } else {
                try {
                    result.success(true)
                } catch (ignored: Exception) {
                }
            }
            "initCamera" -> {
                val hasBarcodeReader = call.argument<Boolean>("hasBarcodeReader")!!
                val flashMode = call.argument<Any>("flashMode").toString()[0]
                val isFillScale = call.argument<Boolean>("isFillScale")!!
                val restrictFormat = call.argument<ArrayList<Int?>>("restrictFormat")!!
                val cameraSelector = call.argument<Int>("cameraSelector")!!
                cameraView!!.initCamera(
                    hasBarcodeReader,
                    flashMode,
                    isFillScale,
                    restrictFormat,
                    cameraSelector
                )
            }
            "resumeCamera" -> cameraView!!.resumeCamera()
            "pauseCamera" -> cameraView!!.pauseCamera()
            "changeFlashMode" -> {
                val captureFlashMode = call.argument<Any>("flashMode").toString()[0]
                cameraView!!.changeFlashMode(captureFlashMode)
            }
            "dispose" -> dispose()
            "setCameraVisible" -> {
                val isCameraVisible = call.argument<Boolean>("isCameraVisible")!!
                cameraView!!.setCameraVisible(isCameraVisible)
            }
            else -> result.notImplemented()
        }
    }

    override fun getView(): View {
        return cameraView!!.view
    }

    override fun dispose() {
        cameraView?.dispose()
    }

    override fun onBarcodesRead(barcodes: List<String?>?) {
        channel.invokeMethod("onBarcodesRead", barcodes)
    }

    companion object {
        private const val REQUEST_CAMERA_PERMISSION = 10001
    }

    init {
        channel.setMethodCallHandler(this)
        if (cameraView == null) {
            cameraView = CameraBaseView(activityPluginBinding.activity, this)
        }
    }
}